src/cljdoc/render/api_searchset.clj

```diff
@@ -0,0 +1,154 @@
+
+(defn path-for-def
+  [version-entity name-of-ns def-name]
+  (str (path-for-namespace version-entity name-of-ns) "#" def-name))
+
+(defn- elements->doc-segments
+  "Does the tricky work of walking through the Jsoup node siblings and collecting
+  headers and their text. Important considerations:
+
+  1. Documents do not necessarily start with a header.
+  2. Headers do not necessarily have text between them and the next header or the
+     end of the document.
+  3. Flexmark makes sure that all markdown headers have anchors, meaning this only works for markdown and not asciidoc."
+  [elements doc-title]
+  (loop [doc-segment {:title doc-title
+                      :anchor nil
+                      :text nil}
+         [element & remaining-elements] elements
+         doc-segments []]
+    (cond
+      (nil? element)
+      (if (:text doc-segment)
+        (conj doc-segments doc-segment)
+        doc-segments)
+
+      (header-tag? (.tagName element))
+      (recur {:title (.text element)
+              :anchor (-> element .children (.select ".md-anchor") .first .attributes (.get "id"))
+              :text nil}
+             remaining-elements
+             (if (:text doc-segment)
+               (conj doc-segments doc-segment)
+               doc-segments))
+
+      :else
+      (recur (update doc-segment :text #(str % " " (.text element)))
+             remaining-elements
+             doc-segments))))
+
+(defn- markdown-doc->docs
+  [doc version-entity]
+  (let [doc-title (:title doc)
+        type-and-contents (doctree/entry->type-and-content doc)
+        html (rt/render-text type-and-contents)
+        doc (Jsoup/parse html)
+        body-elements (-> doc (.getElementsByTag "body") first .children)
+        doc-segments (elements->doc-segments body-elements doc-title)]
+    (map #(let [title (if (:anchor %) (str doc-title " - " (:title %))
+                          doc-title)
+                url (str (path-for-doc doc version-entity) "#" (:anchor %))
+                text (:text %)]
+            {:name title
+             :path url
+             :doc text})
+         doc-segments)))
+
+(defn- generic-doc->docs
+  [doc version-entity]
+  (let [[type contents] (doctree/entry->type-and-content doc)]
+    (when contents
+      (let [html (rt/render-text [type contents])
+            text (-> html Jsoup/parse (.getElementsByTag "body") first .text)]
+        [{:name (:title doc)
+          :path (path-for-doc doc version-entity)
+          :doc text}]))))
+
+(defn- doc->docs
+  "Performs one of two operations, depending on document type:
+
+  1. For markdown it takes a single cache-bundle doc and breaks it into multiple sections by header,
+  each with their own name, doc, and URL path.
+  2. For anything else it renders to HTML and then grabs all the text from the body as a single link."
+  [doc version-entity]
+  (case (:cljdoc.doc/type doc)
+    :cljdoc/markdown (markdown-doc->docs doc version-entity)
+    ;; TODO: asciidoc rendering
```

Comment:
This is still a thing to do but I think we can ship what we have here and add this later.